modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,470 @@
+         (sql-jdbc.execute/set-parameters! driver stmt params)
+         stmt
+         (catch Throwable e
+           (.close stmt)
+           (throw e)))))
+
+
+(defmethod sql-jdbc.execute/statement :presto-jdbc
+  [_ ^Connection conn]
+  ;; and similarly for statement (do not set holdability)
+  (let [stmt (.createStatement conn
+                               ResultSet/TYPE_FORWARD_ONLY
+                               ResultSet/CONCUR_READ_ONLY)]
+       (try
+         (try
+           (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+           (catch Throwable e
+             (log/debug e (trs "Error setting statement fetch direction to FETCH_FORWARD"))))
+         stmt
+         (catch Throwable e
+           (.close stmt)
+           (throw e)))))
+
+(defmethod driver/can-connect? :sql-jdbc
+  [driver details]
+  (sql-jdbc.conn/can-connect? driver details))
+
+(defn- ^PrestoConnection pooled-conn->presto-conn
+  "Unwraps the C3P0 `pooled-conn` and returns the underlying `PrestoConnection` it holds."
+  [^C3P0ProxyConnection pooled-conn]
+  (.unwrap pooled-conn PrestoConnection))
+
+(defmethod sql-jdbc.execute/connection-with-timezone :presto-jdbc
+  [driver database ^String timezone-id]
+  ;; Presto supports setting the session timezone via a `PrestoConnection` instance method. Under the covers,
+  ;; this is equivalent to the `X-Presto-Time-Zone` header in the HTTP request (i.e. the `:presto` driver)
+  (let [conn            (.getConnection (sql-jdbc.execute/datasource database))
+        underlying-conn (pooled-conn->presto-conn conn)]
+    (try
+      (sql-jdbc.execute/set-best-transaction-level! driver conn)
+      (when-not (str/blank? timezone-id)
+        ;; set session time zone if defined
+        (.setTimeZoneId underlying-conn timezone-id))
+      (try
+        (.setReadOnly conn true)
+        (catch Throwable e
+          (log/debug e (trs "Error setting connection to read-only"))))
+      ;; as with statement and prepared-statement, cannot set holdability on the connection level
+      conn
+      (catch Throwable e
+        (.close conn)
+        (throw e)))))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetDateTime]
+  [_ prepared-statement i ^OffsetDateTime t]
+  ;; necessary because `PrestoPreparedStatement` does not support `OffsetDateTime` in its setObject method
+  ;; convert to a SQL `timestamp` at UTC (offset 0), and set that
+  (jdbc/set-parameter (t/sql-timestamp (t/with-offset-same-instant t (t/zone-offset 0))) prepared-statement i))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc ZonedDateTime]
+  [_ prepared-statement i ^OffsetDateTime t]
+  ;; necessary because `PrestoPreparedStatement` does not support `ZonedDateTime` in its setObject method
+  ;; convert to a SQL `timestamp` at UTC, and set that
+  (jdbc/set-parameter (t/sql-timestamp (t/with-zone-same-instant t (t/zone-id "UTC"))) prepared-statement i))
```

Comment:
Is this easier than just converting the temporal objects to temporal literal strings like the stuff we looked at together a couple weeks ago? I'm wary about using `java.sql.Timestamp` since it does weird stuff. I guess it's fine if it works tho

Reply:
OK, I realized that what was needed was to override the `->honeysql` method for `OffsetDateTime` (it was already being done for `ZonedDateTime`).  I have made that change now, and this path is no longer exercised (and the tests pass).

However, is it still a good idea to override these?  If we allow it to fall back to the legacy impl, then the Presto JDBC driver will definitely throw a `NotImplementedException` (since the legacy method will invoke the setTimestamp method taking a `Calendar` as its last parameter, which the Presto JDBC driver does not implement).  So if it's possible there are any code paths that might still attempt to set this parameter (of type `OffsetDateTime`) then we may want to leave it in so they succeed instead of failing?